src/bin/julialauncher.rs

```diff
@@ -179,6 +183,680 @@ fn is_interactive() -> bool {
+/// Returns `None` if no manifest is found (falls back to release channel).
+/// Returns error only if project was specified but couldn't be resolved.
+fn determine_project_version_spec(args: &[String]) -> Result<Option<String>> {
+    determine_project_version_spec_impl(
+        args,
+        std::env::var("JULIA_PROJECT").ok(),
+        std::env::var("JULIA_LOAD_PATH").ok(),
+        &std::env::current_dir().with_context(|| "Failed to determine current directory.")?,
+    )
+}
+
+fn determine_project_version_spec_impl(
+    args: &[String],
+    julia_project: Option<String>,
+    julia_load_path: Option<String>,
+    current_dir: &Path,
+) -> Result<Option<String>> {
+    let depot_path = std::env::var_os("JULIA_DEPOT_PATH");
+
+    // Get the active project file using Julia's init_active_project logic
+    let project_file = match init_active_project_impl(
+        args,
+        current_dir,
+        julia_project.as_deref(),
+        depot_path.as_deref(),
+    )? {
+        Some(file) => file,
+        None => {
+            // If no project specified via --project or JULIA_PROJECT,
+            // try searching JULIA_LOAD_PATH
+            if let Some(load_path) = julia_load_path {
+                match find_project_from_load_path(&load_path, current_dir, depot_path.as_deref())? {
+                    Some(file) => file,
+                    None => {
+                        log::debug!("VersionDetect::No project specification found");
+                        return Ok(None);
+                    }
+                }
+            } else {
+                log::debug!("VersionDetect::No project specification found");
+                return Ok(None);
+            }
+        }
+    };
+
+    log::debug!(
+        "VersionDetect::Using project file: {}",
+        project_file.display()
+    );
+
+    // Find the manifest file using Julia's project_file_manifest_path logic
+    let manifest_path = match project_file_manifest_path(&project_file) {
+        Some(path) => path,
+        None => {
+            log::debug!("VersionDetect::No manifest file found for project");
+            return Ok(None);
+        }
+    };
+
+    log::debug!(
+        "VersionDetect::Detected manifest file: {}",
+        manifest_path.display()
+    );
+
+    // Read julia_version from manifest
+    if let Some(version) = read_manifest_julia_version(&manifest_path)? {
+        log::debug!(
+            "VersionDetect::Read Julia version from manifest: {}",
+            version
+        );
+        return Ok(Some(version));
+    }
+
+    log::debug!("VersionDetect::Manifest file exists but does not contain julia_version field");
+    Ok(None)
+}
+
+/// Find highest versioned manifest, preferring JuliaManifest over Manifest for same version
+fn find_highest_versioned_manifest(project_root: &Path) -> Option<PathBuf> {
+    let Ok(entries) = fs::read_dir(project_root) else {
+        return None;
+    };
```

Comment:
Does this read all files in the directory? Maybe we could just filter matching files, and then do a single reduction operation on top of that? It might be easier to read too